### PR TITLE
Code inspection

### PR DESCRIPTION
**Pull request by catena2w**

### CODE REVIEW

File: src/main/scala/org/ergoplatform/modifiers/history/Header.scala [hunk 1]
```diff
@@ -151,7 +155,7 @@ object HeaderSerializer extends Serializer[Header] {
     val stateRoot = ADDigest @@ bytes.slice(97, 130)
     val timestamp = Longs.fromByteArray(bytes.slice(130, 138))
     val votes = bytes.slice(138, 143)
-    val nBits = RequiredDifficulty.parseBytes(bytes.slice(143, 147)).get
+    val nBits = RequiredDifficulty.parseBytes(bytes.slice(143, 147)).getOrElse(throw new IllegalArgumentException("cannot parse NBITS"))
```
Comment (kushti) [inline src/main/scala/org/ergoplatform/modifiers/history/Header.scala]: .getOrElse(throw new) is not any different from .get and breaks function signature "getOrElse[B >: A](default: => B): B"

File: src/main/scala/org/ergoplatform/modifiers/history/PoPoWProof.scala [hunk 2]
```diff
@@ -119,13 +122,15 @@ class PoPoWProofSerializer(poWScheme: PoWScheme) extends Serializer[PoPoWProof]
     val k = bytes(1)
     val i = bytes(2)
     val headSuffixLength = Shorts.fromByteArray(bytes.slice(3, 5))
-    val headSuffix = HeaderSerializer.parseBytes(bytes.slice(5, 5 + headSuffixLength)).get
+    val headSuffix = HeaderSerializer.parseBytes(bytes.slice(5, 5 + headSuffixLength))
+      .getOrElse(throw new IllegalArgumentException("Cannot parse bytes to header"))
```
Comment (kushti) [inline src/main/scala/org/ergoplatform/modifiers/history/PoPoWProof.scala]: .getOrElse(exception)

File: src/main/scala/org/ergoplatform/nodeView/history/ErgoHistoryReader.scala [hunk 3]
```diff
@@ -157,27 +158,34 @@ trait ErgoHistoryReader
...
+      else {
+        val updatedChains = nextLevelHeaders
+          .flatMap(h => acc.find(chain => h.parentId sameElements chain.head.id).map(c => h +: c))
+        val nonUpdatedChains = acc.filter(chain => !nextLevelHeaders.exists(_.parentId sameElements chain.head.id))
+        loop(currentHeight + 1, updatedChains ++ nonUpdatedChains)
+      }
     }
 
-    loop(Seq(header))
+    loop(heightOf(header.id).get, Seq(Seq(header)))
```
Comment (kushti) [inline src/main/scala/org/ergoplatform/nodeView/history/ErgoHistoryReader.scala]: .get

File: src/main/scala/org/ergoplatform/modifiers/history/PoPoWProof.scala [hunk 4]
```diff
@@ -122,10 +126,12 @@ class PoPoWProofSerializer(poWScheme: PoWScheme) extends Serializer[PoPoWProof]
...
     def parseSuffixes(index: Int, acc: Seq[Header]): (Int, Seq[Header]) = {
-      if (acc.length == k) (index, acc.reverse)
-      else {
+      if (acc.lengthCompare(k.toInt) == 0)  {
+        (index, acc.reverse)
+      } else {
         val l = Shorts.fromByteArray(bytes.slice(index, index + 2))
-        val headerWithoutInterlinks = HeaderSerializer.parseBytes(bytes.slice(index + 2, index + 2 + l)).get
+        val headerWithoutInterlinks = HeaderSerializer.parseBytes(bytes.slice(index + 2, index + 2 + l))
+          .getOrElse(throw new IllegalArgumentException("Cannot parse bytes to header"))
```
Comment (kushti) [inline src/main/scala/org/ergoplatform/modifiers/history/PoPoWProof.scala]: .getOrElse(throw()) !

File: src/main/scala/org/ergoplatform/modifiers/history/PoPoWProof.scala [hunk 5]
```diff
@@ -136,10 +142,11 @@ class PoPoWProofSerializer(poWScheme: PoWScheme) extends Serializer[PoPoWProof]
     index = index + 2
     val innerchain = (0 until innerchainLength) map { _ =>
       val l = Shorts.fromByteArray(bytes.slice(index, index + 2))
-      val header = HeaderSerializer.parseBytes(bytes.slice(index + 2, index + 2 + l)).get
+      val header = HeaderSerializer.parseBytes(bytes.slice(index + 2, index + 2 + l))
+        .getOrElse(throw new IllegalArgumentException("Cannot parse bytes to header"))
```
Comment (kushti) [inline src/main/scala/org/ergoplatform/modifiers/history/PoPoWProof.scala]: .getOrElse(throw()) !

File: src/main/scala/org/ergoplatform/nodeView/history/ErgoHistoryReader.scala (hunk 3 above)
Comment (daron666) [reply src/main/scala/org/ergoplatform/nodeView/history/ErgoHistoryReader.scala]: @kushti Issue has been created #138

**Pull request merged**